README.md
# flutter_image_pick_crop

A new Flutter plugin for cropping images picked from camera or gallery.

## Screenshots

<img src="ss1.png" height="300em" /> <img src="ss2.png" height="300em" />
<img src="ss3.png" height="300em" /> <img src="ss4.png" height="300em" />
<img src="ss5.png" height="300em" /> <img src="ss6.png" height="300em" />

### Show some :heart: and star the repo to support the project

## Usage

[Example](https://github.com/mtellect/FlutterImagePickCrop/blob/master/example/lib/main.dart)

```yaml
  dependencies:
    flutter:
      sdk: flutter
    flutter_image_pick_crop:
```

```dart
class _MyAppState extends State<MyApp> {
  String _platformMessage = 'Unknown';
  String _camera = 'fromCameraCropImage';
  String _gallery = 'fromGalleryCropImage';
  File imageFile;

  @override
  initState() {
    super.initState();
  }


  initGalleryPickUp() async {
    /*setState(() {
      imageFile = null;
      _platformVersion = 'Unknown';
    });*/
    File file;
    String result;

    try {
      result = await FlutterImagePickCrop.pickAndCropImage(_gallery);
    } on PlatformException catch (e) {
      result = e.message;
      print(e.message);
    }

    if (!mounted) return;

    setState(() {
      imageFile = new File(result);
      _platformMessage = result;
    });
  }

  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: new Text('Plugin Image Crop'),
        ),
        body: new Container(
          padding: const EdgeInsets.all(8.0),
          child: new Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              imageFile != null
                  ? new Image.file(
                      imageFile,
                      height: 400.0,
                    )
                  : new Icon(
                      Icons.image,
                      size: 200.0,
                      color: Theme.of(context).primaryColor,
                    ),
              new Center(
                child: new Text('Running on: $_platformMessage\n'),
              ),
              new RaisedButton.icon(
                  onPressed: initGalleryPickUp,
                  icon: new Icon(Icons.image),
                  label: new Text("Open Camera/Gallery")),
              new Padding(padding: new EdgeInsets.only(left: 5.0, right: 5.0))
            ],
          ),
        ),
      ),
    );
  }
}
```


# Pull Requests

I welcome and encourage all pull requests. It usually will take me within 24-48 hours to respond to any issue or request. Here are some basic rules to follow to ensure timely addition of your request:

1.  Match coding style (braces, spacing, etc.) This is best achieved using `Reformat Code` feature of Android Studio `CMD`+`Option`+`L` on Mac and `CTRL` + `ALT` + `L` on Linux + Windows .
2.  If its a feature, bugfix, or anything please only change code to what you specify.
3.  Please keep PR titles easy to read and descriptive of changes, this will make them easier to merge :)
4.  Pull requests _must_ be made against `develop` branch. Any other branch (unless specified by the maintainers) will get rejected.
5.  Check for existing [issues](https://github.com/mtellect/FlutterImagePickCrop/issues) first, before filing an issue.
6.  Make sure you follow the set standard as all other projects in this repo do
7.  Have fun!

### Created & Maintained By

[Maugost_Mtellect](https://github.com/mtellect/)

> If you found this project helpful or you learned something from the source code and want to thank me, consider buying me a cup of :coffee:
>
> *Bitcoin: 1K2wivc2xy7tTjsmHhNzq7wwEHoyHfS64n


# License

    Copyright 2018 Maugost Mtellect

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


## Getting Started

For help getting started with Flutter, view our online [documentation](https://flutter.io/).

For help on editing package code, view the [documentation](https://flutter.io/developing-packages/).

CHANGELOG.md
## [0.0.1] - 6/62018.

* Now you can pick images from either camera or gallery and crop the images before being displayed. (This version currently supports android for now.)


example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_image_pick_crop_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());

    // Verify that platform version is retrieved.
    expect(
        find.byWidgetPredicate(
          (Widget widget) =>
              widget is Text && widget.data.startsWith('Running on:'),
        ),
        findsOneWidget);
  });
}


android/src/main/java/maugost/flutterimagepickcrop/FlutterImagePickCropPlugin.java
package maugost.flutterimagepickcrop;

import android.Manifest;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.provider.MediaStore;
import android.support.annotation.RequiresApi;
import android.view.View;
import android.widget.Toast;

import com.myhexaville.smartimagepicker.ImagePicker;
import com.theartofdev.edmodo.cropper.CropImage;
import com.theartofdev.edmodo.cropper.CropImageView;

import java.io.File;
import java.net.URI;
import java.net.URISyntaxException;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import static android.app.Activity.RESULT_OK;

/**
 * FlutterImagePickCropPlugin
 */
public class FlutterImagePickCropPlugin implements MethodCallHandler, PluginRegistry.ActivityResultListener, PluginRegistry.RequestPermissionsResultListener {
    private static final String PICK_OPTIONS = "Options";
    private static final String REQUEST_FOR_GALLERY = "fromGalleryCropImage";
    private static final String REQUEST_FOR_CAMERA = "fromCameraCropImage";
    private static final int PICK_FROM_CAMERA = 1000;
    private static final int PICK_FROM_GALLARY = 2000;
    private static final String PICK_AND_CROP = "pickAndCropImage";
    private final MethodChannel channel;
    private ImagePicker imagePicker;
    private Activity activity;
    private Context context;
    private Uri destinationUri;
    private Uri mCropImageUri;
    private Result pendingResult;
    private MethodCall methodCall;


    private FlutterImagePickCropPlugin(Activity activity, Context context, MethodChannel channel) {
        this.activity = activity;
        this.context = context;
        this.channel = channel;
    }

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_image_pick_crop");

        FlutterImagePickCropPlugin instance = new FlutterImagePickCropPlugin(registrar.activity(), registrar.context(), channel);

        registrar.addRequestPermissionsResultListener(instance);
        registrar.addActivityResultListener(instance);
        channel.setMethodCallHandler(instance);

    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        /*if ("getPlatformVersion".equals(call.method)) {
            pendingResult.success("Android " + Build.VERSION.RELEASE);
        } else*/

        if (!setPendingMethodCallAndResult(methodCall, result)) {
            finishWithAlreadyActiveError();
            return;
        }

        if (PICK_AND_CROP.equals(call.method)) {
            checkRequestResult(call);
        } else {
            pendingResult.notImplemented();
        }
    }

    private void checkRequestResult(MethodCall call) {
        if (call.hasArgument(PICK_OPTIONS)) {
            String options = call.argument(PICK_OPTIONS).toString();

            //Toast.makeText(activity, "Option Received", Toast.LENGTH_SHORT).show();

            if (options.equals(REQUEST_FOR_CAMERA)) {
                openPickupFromCamera();
            } else if (options.equals(REQUEST_FOR_GALLERY)) {
                openPickupFromGallery();
            }
        }
    }

    private void openPickupFromCamera() {
        //Toast.makeText(activity, REQUEST_FOR_CAMERA, Toast.LENGTH_SHORT).show();

        /* For Image capture from camera */
        Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        activity.startActivityForResult(intent, PICK_FROM_CAMERA);


    }

    private void openPickupFromGallery() {
        //Toast.makeText(activity, REQUEST_FOR_GALLERY, Toast.LENGTH_SHORT).show();

        // For Image capture from Gallary
   /*     activity.startActivityForResult(new Intent(Intent.ACTION_PICK, android.provider.MediaStore.Images.Media.INTERNAL_CONTENT_URI),
                PICK_FROM_GALLARY);*/
        CropImage.startPickImageActivity(activity);

    }


    private void startCropImageActivity() {
        CropImage.activity()
                .start(activity);
    }


    public void onSelectImageClick(View view) {
        CropImage.startPickImageActivity(activity);
    }

    private void startCropImageActivity(Uri imageUri) {
        CropImage.activity(imageUri).setGuidelines(CropImageView.Guidelines.ON)
                .start(activity);


    }

    @TargetApi(Build.VERSION_CODES.M)
    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {

        if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            CropImage.ActivityResult result = CropImage.getActivityResult(data);
            if (resultCode == RESULT_OK) {
                Uri resultUri = result.getUri();
                finishWithSuccess(resultUri.getPath());
            } else if (resultCode == CropImage.CROP_IMAGE_ACTIVITY_RESULT_ERROR_CODE) {
                Exception error = result.getError();
                finishWithError("Failed Permission",error.getMessage());
            }

            return true;
        }

    if (requestCode == CropImage.PICK_IMAGE_CHOOSER_REQUEST_CODE && resultCode == RESULT_OK) {
            Uri imageUri = CropImage.getPickImageResultUri(context, data);
            if (CropImage.isReadExternalStoragePermissionsRequired(context, imageUri)) {
                mCropImageUri = imageUri;
                //Toast.makeText(activity, "check: " + mCropImageUri.toString(), Toast.LENGTH_SHORT).show();
                activity.requestPermissions(new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, CropImage.PICK_IMAGE_PERMISSIONS_REQUEST_CODE);
            } else {
                startCropImageActivity(imageUri);
            }


            return true;
        }
//
//        else if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE && resultCode == RESULT_OK) {
//            Uri imageUri = CropImage.getPickImageResultUri(context, data);
//            if (imageUri != null) {
//                finishWithSuccess(imageUri.getPath());
//            }else {
//                finishWithSuccess(null);
//            }
//            return true;
//        }

        return false;
    }


    @Override
    public boolean onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {

        if (requestCode == CropImage.PICK_IMAGE_PERMISSIONS_REQUEST_CODE) {
            if (mCropImageUri != null && grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // required permissions granted, start crop image activity
                startCropImageActivity(mCropImageUri);
            } else {
                finishWithError("Failed Permission","Cancelling, required permissions are not granted");
                //Toast.makeText(activity, "Cancelling, required permissions are not granted", Toast.LENGTH_LONG).show();
            }

            return true;
        }

        return false;
    }

    private void finishWithSuccess(String imagePath) {
        pendingResult.success(imagePath);
        clearMethodCallAndResult();
    }

    private void finishWithAlreadyActiveError() {
        finishWithError("already_active", "Image picker is already active");
    }

    private void finishWithError(String errorCode, String errorMessage) {
        pendingResult.error(errorCode, errorMessage, null);
        clearMethodCallAndResult();
    }

    private void clearMethodCallAndResult() {
        methodCall = null;
        pendingResult = null;
    }

    private boolean setPendingMethodCallAndResult(
            MethodCall methodCall, MethodChannel.Result result) {
        if (pendingResult != null) {
            return false;
        }

        this.methodCall = methodCall;
        pendingResult = result;
        return true;
    }
}


example/lib/main.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_image_pick_crop/flutter_image_pick_crop.dart';

void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => new _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformMessage = 'Unknown';
  String _camera = 'fromCameraCropImage';
  String _gallery = 'fromGalleryCropImage';
  File imageFile;

  @override
  initState() {
    super.initState();
  }

  initGalleryPickUp() async {
    /*setState(() {
      imageFile = null;
      _platformVersion = 'Unknown';
    });*/
    File file;
    String result;

    try {
      result = await FlutterImagePickCrop.pickAndCropImage(_gallery);
    } on PlatformException catch (e) {
      result = e.message;
      print(e.message);
    }

    if (!mounted) return;

    setState(() {
      imageFile = new File(result);
      _platformMessage = result;
    });
  }

  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: new Text('Plugin Image Crop'),
        ),
        body: new Container(
          padding: const EdgeInsets.all(8.0),
          child: new Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              imageFile != null
                  ? new Image.file(
                      imageFile,
                      height: 400.0,
                    )
                  : new Icon(
                      Icons.image,
                      size: 200.0,
                      color: Theme.of(context).primaryColor,
                    ),
              new Center(
                child: new Text('Running on: $_platformMessage\n'),
              ),
              new RaisedButton.icon(
                  onPressed: initGalleryPickUp,
                  icon: new Icon(Icons.image),
                  label: new Text("Open Camera/Gallery")),
              new Padding(padding: new EdgeInsets.only(left: 5.0, right: 5.0))
            ],
          ),
        ),
      ),
    );
  }
}
